src/odemis/driver/rigol.py

```diff
@@ -0,0 +1,438 @@
+        err_code, err_msg = self.QueryErrorState()
+        if err_code is not None:
+            err = "Error code %d from %s: %s" % (err_code, self.name, err_msg)
+            logging.error(err)
+            raise model.HwError(err)
+    
+    @classmethod
+    def _openConnection(cls, host, port):
+        """
+        Open a TCP/IP connection with the device
+        return (Accesser)
+        """
+        return IPAccesser(host, port)
+    
+
+class IPAccesser(object):
+    """
+    Manages low-level connections over IP
+    """
+    def __init__(self, host, port=DEFAULT_PORT):
+        """
+        host (string): the IP address or host name of the master controller
+        port (int): the (IP) port number
+        """
+        self._host = host
+        self._port = port
+        self._is_connected = False
+        
+        if self._host == "fake":
+            self.simulator = FakeDG1000Z()
+            self._host = "localhost"
+        else:
+            self.simulator = None
+            
+        try:
+            logging.debug("Connecting to %s:%d" % (self._host, self._port))
+            self.socket = socket.create_connection((self._host, self._port), timeout=1)
+            self._is_connected = True
+        except socket.error:
+            raise model.HwError("Failed to connect to '%s:%d', check that the Rigol "
+                                "Clock Generator is connected to the network, turned "
+                                "on, and correctly configured." % (host, port))
+
+        # to acquire before sending anything on the socket
+        self._net_access = threading.Lock()
+        
+    def __del__(self):
+        self.terminate()
+        
+    def terminate(self):
+        if self.simulator:
+            self.simulator.terminate()
+        if self._is_connected:
+            self.socket.close()
+        self._is_connected = False
+
+    def sendOrderCommand(self, cmd, val=""):
+        """
+        Sends one command, and don't expect any reply
+        cmd (str): command to send
+        val (str): value to send (if any)
+        raises:
+            IOError: if problem with sending/receiving data over the connection
+        """
+        if not self._is_connected:
+            raise IOError("Device %s not connected." % (self._host,))
+
+        msg = "%s %s\n" % (cmd, val)
+
+        with self._net_access:
+            self.socket.sendall(msg)
+
+    def sendQueryCommand(self, cmd):
+        """
+        Sends one command, and don't expect any reply
+        cmd (str): command to send, including the ?
+
+        raises:
+            IOError: if problem with sending/receiving data over the connection
+        """
+        if not self._is_connected:
+            raise IOError("Device %s not connected." % (self._host))
+        
+        msg = "%s\n" % cmd
+
+        with self._net_access:
+            self.socket.sendall(msg)
+
+            data = None
+            try:
+                data = self.socket.recv(4096)
```

Comment:
This is a really (too) simple implementation of the socket. In theory, a responce might be splitted over multiple packets. So you need to wait until you receive a line termination character. See pigcs driver or jeol (form line 263).